### PR TITLE
bpo-29779: new environment variable PYTHONHISTORY.

### PR DESCRIPTION
**Pull request by ghost**
if this variable is set the user can change the location of
a ~/.python_history file without adding hook to PYTHONSTARTUP.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Comment by ghost** [comment]
I signed the CLA :wink:

**Comment by warsaw** [comment]
This is a nice change, and it would probably let me get rid of my $PYTHONSTARTUP setting.  I wonder if `python3 -h` should mention this new envar?

**Comment by ghost** [comment]
@warsaw I added this to Misc/python.man

**Comment by warsaw** [comment]
@0xl3vi Thanks, that looks good.  It's still not mentioned in `python3 -h` but I'm not entirely sure it needs to.  There are other `PYTHON*` environment variables not mentioned in the `--help` output.

File: Lib/site.py
Comment (warsaw) [inline Lib/site.py]: Minor style nit: You don't need this extra blank line.

**Review by warsaw — APPROVED** [review]
LGTM, with one very minor style nit.

**Comment by ghost** [comment]
@warsaw I updated main.c for `python -h`,
blank line fixed.

**Comment by warsaw** [comment]
The changes look great to me, but I'm going to wait just a bit before merging this.

* I'd like to see if any other core dev has an opinion about the change
* The AppVeyor test is failing, although I suspect that is unrelated to your change

I don't have a Windows environment to test that on.  Maybe @zooba can weigh in on that.  Alternatively, try rebasing against master if it's been fixed there.

If there are no other follow ups by this weekend, I'll merge your change.

Can you please also add a `Misc/NEWS` entry (knowing that our workflow for that kind of sucks and we might get conflicts)?

**Review by ned-deily — CHANGES_REQUESTED** [review]
Environment variables are also documented in the Python Setup and Usage document (https://docs.python.org/3/using/cmdline.html#environment-variables) which is generated from Doc/using/cmdline.rst.

**Comment by ned-deily** [comment]
Also there should be an issue opened on bugs.python.org to document this and the issue number added to the title of this pull request.

**Comment by ghost** [comment]
@warsaw I tried rebasing against master and the build passed!
and Misc/NEWS updated.

@ned-deily I updated cmdline.rst and created new issue http://bugs.python.org/issue29779

P.S: I need to add bpo-XXX or just the issue number in bugs.python.org?

File: Lib/site.py
Comment (warsaw) [inline Lib/site.py]: You could also do something like:

```python
    return os.environ.get(
        'PYTHONHISTORY',
        os.path.expanduser(os.path.join('~', '.python_history')))
```

Also `set_history_file()` is a little bit of a misnomer since it's not actually setting it, but calculating it instead.  OTOH, I bet the more compact form could just be inlined below.

File: Misc/NEWS
Comment (warsaw) [inline Misc/NEWS]: Change `Issue #29779` to `bpo-29779`.  That's the new style of marking issue numbers.

**Review by warsaw — CHANGES_REQUESTED** [review]
Just a couple of other thoughts on the implementation and NEWS file entry.

**Comment by warsaw** [comment]
Oh, and it would be nice to add a test for codecov.

**Comment by ned-deily** [comment]
Thanks for addressing my comments!

**Review by ned-deily — APPROVED** [review]

**Comment by ghost** [comment]
@warsaw all done, I changed set_history_file() to gethistoryfile().
and added test.

**Comment by warsaw** [comment]
@0xl3vi We still need a test.

**Comment by warsaw** [comment]
Oh and yay for Misc/NEWS conflicts. :(

**Comment by warsaw** [comment]
Also note [this report](http://bugs.python.org/issue29779#msg289497) which I can confirm.

**Comment by ghost** [comment]
@warsaw sorry for the dealy.
what tests we need add?

regarding [this](http://bugs.python.org/issue29779#msg289497)
If the PYTHONHISTORY is empty it will fall back to .python_history.
there is no error on my build

**Comment by warsaw** [comment]
@0xl3vi At the very least, I think a test covering `gethistoryfile()` would be useful.   If you can mock out enough of the readline stuff (or other safeguards so as not to interfere with the environment of a person running the tests), then adding some coverage for `enablerlcompleter()` would be nice.

The bug happens when the file pointed to by `PYTHONHISTORY` does not exist.

**Comment by ghost** [comment]
@warsaw OK updated.

* In case PYTHONHISTORY will be empty or not set,
it wil fall back to the default history file.
if the file pointed by PYTHONHISTORY does not exist readline will create one.

The test checks if gethistoryfile() reads the variable.
Bash will expand '~' so we dont need to to test the path.

what do you think?

File: Doc/using/cmdline.rst
Comment (berkerpeksag) [inline Doc/using/cmdline.rst:719]: \`\`~/.python_history\`\`

File: Doc/using/cmdline.rst
Comment (berkerpeksag) [inline Doc/using/cmdline.rst:718]: "a \`\`.python_history\`\` file" or just "a history file".

File: Lib/site.py
Comment (berkerpeksag) [inline Lib/site.py:381]: What if `os.environ.get("PYTHONHISTORY")` returns `None`?

File: Lib/test/test_site.py
Comment (berkerpeksag) [inline Lib/test/test_site.py:263]: Please use `EnvironmentVarGuard` in `test.support` to set env variables.

File: Misc/NEWS
Comment (berkerpeksag) [inline Misc/NEWS:14]: Please add "Patch by Your Name."

File: Misc/python.man
Comment (berkerpeksag) [inline Misc/python.man:436]: File names should be italic in man pages so `\fI~/.python_history\fI` or

```man
python_history file, by default it will be
.IR ~/.python_history .
```

You may need to escape ~ or / characters.

File: Modules/main.c
Comment (berkerpeksag) [inline Modules/main.c:109]: This line is a bit long.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]
This also needs to be documented in `Doc/whatsnew/3.7.rst` (and please add "(Contributed by Your Name.)")

Thanks!

**Comment by ghost** [comment]
OK, I don't have time to do it.
If someone else want to implement this please do so.

**Comment by warsaw** [comment]
@0xl3vi Hi.  I'm sorry this has taken so long.  Are you saying in your comment that you don't have time to finish this branch?  If so, that's okay and we greatly appreciate your contribution so far.  I'm willing to finish this branch for you if you'd like.

**Comment by ghost** [comment]
@warsaw sure if you want.
thanks :+1:

**Comment by ZackerySpytz** [comment]
@warsaw @berkerpeksag I have created a new PR for this feature: GH-13208. It addresses all of @berkerpeksag's comments.

**Pull request closed without merging**
